File: pdf_ocr.swift
import Foundation
import AppKit
import PDFKit
import Vision

struct OCRResult {
    let pageIndex: Int
    let text: String
}

func renderPage(_ page: PDFPage, scale: CGFloat) -> CGImage? {
    let bounds = page.bounds(for: .mediaBox)
    let width = Int(bounds.width * scale)
    let height = Int(bounds.height * scale)
    guard let colorSpace = CGColorSpace(name: CGColorSpace.sRGB),
          let context = CGContext(
            data: nil,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: 0,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
          ) else {
        return nil
    }

    context.setFillColor(NSColor.white.cgColor)
    context.fill(CGRect(x: 0, y: 0, width: width, height: height))
    context.saveGState()
    context.translateBy(x: 0, y: CGFloat(height))
    context.scaleBy(x: scale, y: -scale)
    page.draw(with: .mediaBox, to: context)
    context.restoreGState()

    return context.makeImage()
}

func recognizeText(from image: CGImage) throws -> String {
    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = false
    request.recognitionLanguages = ["ja-JP", "en-US"]
    request.minimumTextHeight = 0.005

    let handler = VNImageRequestHandler(cgImage: image, options: [:])
    try handler.perform([request])

    let observations = request.results ?? []
    let ordered = observations.sorted {
        if abs($0.boundingBox.midY - $1.boundingBox.midY) > 0.01 {
            return $0.boundingBox.midY > $1.boundingBox.midY
        }
        return $0.boundingBox.minX < $1.boundingBox.minX
    }

    return ordered.compactMap { $0.topCandidates(1).first?.string }.joined(separator: "\n")
}

func main() throws {
    guard CommandLine.arguments.count >= 2 else {
        fputs("usage: swift pdf_ocr.swift <pdf-path> [scale]\n", stderr)
        exit(2)
    }

    let pdfPath = CommandLine.arguments[1]
    let scale = CommandLine.arguments.count >= 3 ? CGFloat(Double(CommandLine.arguments[2]) ?? 2.0) : 2.0

    guard let document = PDFDocument(url: URL(fileURLWithPath: pdfPath)) else {
        fputs("failed to open pdf: \(pdfPath)\n", stderr)
        exit(1)
    }

    var results: [OCRResult] = []
    for pageIndex in 0..<document.pageCount {
        guard let page = document.page(at: pageIndex),
              let image = renderPage(page, scale: scale) else {
            continue
        }
        let text = try recognizeText(from: image)
        results.append(OCRResult(pageIndex: pageIndex + 1, text: text))
    }

    for result in results {
        print("=== PAGE \(result.pageIndex) ===")
        print(result.text)
    }
}

try main()

File: pdf_ocr.py
import sys

import fitz
import Foundation
import Quartz
import Vision


def png_bytes_to_cgimage(png_bytes):
    data = Foundation.NSData.dataWithBytes_length_(png_bytes, len(png_bytes))
    source = Quartz.CGImageSourceCreateWithData(data, None)
    if source is None:
        raise RuntimeError("failed to create image source")
    image = Quartz.CGImageSourceCreateImageAtIndex(source, 0, None)
    if image is None:
        raise RuntimeError("failed to decode PNG to CGImage")
    return image


def ocr_cgimage(image):
    request = Vision.VNRecognizeTextRequest.alloc().init()
    request.setRecognitionLevel_(Vision.VNRequestTextRecognitionLevelAccurate)
    request.setUsesLanguageCorrection_(False)
    request.setRecognitionLanguages_(["ja-JP", "en-US"])

    handler = Vision.VNImageRequestHandler.alloc().initWithCGImage_options_(image, None)
    ok, error = handler.performRequests_error_([request], None)
    if not ok:
        raise RuntimeError(str(error))

    results = request.results() or []
    items = []
    for obs in results:
        candidates = obs.topCandidates_(1)
        if not candidates:
            continue
        box = obs.boundingBox()
        items.append(
            {
                "text": str(candidates[0].string()),
                "x": box.origin.x,
                "y": box.origin.y,
                "h": box.size.height,
            }
        )

    items.sort(key=lambda item: (-item["y"], item["x"]))

    rows = []
    threshold = 0.012
    for item in items:
        if not rows or abs(rows[-1]["y"] - item["y"]) > threshold:
            rows.append({"y": item["y"], "items": [item]})
        else:
            rows[-1]["items"].append(item)

    lines = []
    for row in rows:
        row["items"].sort(key=lambda item: item["x"])
        line = "\t".join(item["text"] for item in row["items"])
        lines.append(line)
    return "\n".join(lines)


def ocr_pdf(pdf_path, scale):
    doc = fitz.open(pdf_path)
    for page_number, page in enumerate(doc, start=1):
        pix = page.get_pixmap(matrix=fitz.Matrix(scale, scale), alpha=False)
        image = png_bytes_to_cgimage(pix.tobytes("png"))
        print(f"=== PAGE {page_number} ===")
        print(ocr_cgimage(image))


def main():
    if len(sys.argv) < 2:
        print("usage: python3 pdf_ocr.py <pdf-path> [scale]", file=sys.stderr)
        raise SystemExit(2)
    pdf_path = sys.argv[1]
    scale = float(sys.argv[2]) if len(sys.argv) >= 3 else 2.0
    ocr_pdf(pdf_path, scale)


if __name__ == "__main__":
    main()
